Lib/test/test_asyncio/test_selector_thread.py

```diff
@@ -0,0 +1,128 @@
+import asyncio
+import select
+import socket
+import time
+import unittest
+from asyncio._selector_thread import SelectorThread
+from test import support
+from unittest import mock
+
+
+def tearDownModule():
+    asyncio.events._set_event_loop_policy(None)
+
+
+class SelectorThreadTest(unittest.IsolatedAsyncioTestCase):
+    async def asyncSetUp(self):
+        self._sockets = []
+        self.selector_thread = SelectorThread(asyncio.get_running_loop())
+
+    def socketpair(self):
+        pair = socket.socketpair()
+        self._sockets.extend(pair)
+        return pair
+
+    async def asyncTearDown(self):
+        self.selector_thread.close()
+        for s in self._sockets:
+            s.close()
+
+    async def test_slow_reader(self):
+        a, b = self.socketpair()
+        first_recv = asyncio.Future()
+
+        def recv():
+            msg = b.recv(100)
+            if not first_recv.done():
+                first_recv.set_result(msg)
+
+        mock_recv = mock.MagicMock(wraps=recv)
+        # make sure select is only called once when
+        # event loop thread is slow to consume events
+        a.sendall(b"msg")
+        with mock.patch("select.select", wraps=select.select) as mock_select:
+            self.selector_thread.add_reader(b, mock_recv)
+            # ready event, but main event loop is blocked for some time
+            time.sleep(0.1)
```

Comment:
Is this sleep really needed? If yes, can it be replaced with a synchronization primitive instead of a sleep?

Reply:
I would love help writing tests to verify that things _don't_ happen in multithreaded code.

This was added to demonstrate an answer to the question in review that the select loop doesn't keep firing away if the main loop is slow to process one event, and was the best I could come up with to demonstrate that it really is waiting and only calls select once.

The sequence of events that we want to wait for is that the select thread is waiting in select_cond.wait() after select is called the first time. Maybe there's a way to instrument that.